synapseclient/retry.py

```diff
@@ -45,51 +46,48 @@ def foo(a, b, c): return [a, b, c]
             response = function()
         except Exception as ex:
             exc_info = sys.exc_info()
-            if verbose:
-                sys.stderr.write(str(ex.message)+'\n')
+            log_error(str(ex), verbose)
             if hasattr(ex, 'response'):
                 response = ex.response
 
         # Check if we got a retry-able error
         if response is not None:
             if response.status_code in retry_status_codes:
-                if verbose:
-                    sys.stderr.write("retrying on status code: \n" + str(response.status_code))
+                response_message = _get_message(response)
                 retry = True
-
+                log_error("retrying on status code: %s" % str(response.status_code), verbose)
+                log_error(str(response_message))
+                if (response.status_code == 429) and (wait>10):
+                    sys.stderr.write('%s...\n' % response_message)
+                    sys.stderr.write('Retrying in %i seconds' %wait)
+                
             elif response.status_code not in range(200,299):
-                ## if the response is JSON, look for retryable errors in the 'reason' field
-                if _is_json(response.headers.get('content-type', None)):
-                    try:
-                        json = response.json()
-                        ## special case for message throttling
-                        if 'Please slow down.  You may send a maximum of 10 message' in json.get('reason', None):
-                            if verbose: sys.stderr.write("retrying"+ json.get('reason', None)+'\n')
-                            retry = True
-                            wait = 16
-                        elif any([msg.lower() in json.get('reason', None).lower() for msg in retry_errors]):
-                            if verbose: sys.stderr.write("retrying"+ json.get('reason', None)+'\n')
-                            retry = True
-                    except (AttributeError, ValueError) as ex:
-                        pass
-
-                ## if the response is not JSON, look for retryable errors in its text content
-                elif any([msg.lower() in response.text.lower() for msg in retry_errors]):
-                    if verbose: sys.stderr.write("retrying"+ response.text+'\n')
+                ## For all other non 200 messages look for retryable errors in the body or reason field
+                response_message = _get_message(response)
+                if any([msg.lower() in response_message.lower() for msg in retry_errors]):
+                    retry = True
+                    log_error('retrying %s' %response_message, verbose)
+                ## special case for message throttling
+                elif 'Please slow down.  You may send a maximum of 10 message' in response:
```

Comment:
consider getting rid of this case since, (1) generic exp back off may handle it, (2) we shouldn't have logic that switches based on the text of an error message